hashintel/hash
# Make all links a collection of links

## 🌟 What is the purpose of this PR?



This PR makes it so that all links are considered a collection of links (i.e. a list) in the Graph API. This change is to simplify the logic around links and make handling them more consistent. This also aligns better with how structural queries on links work. We now always return lists of links from link queries, and as such, we don't consider "singular links" relevant for queries.

⚠️ This PR will break local dev environments ⚠️

In order to fix local dev environments, it's required to wipe the database completely and rebuild the docker images:

**When running through external-services**:
1. Be in the root yarn workspace, stop the containers: `yarn external-services down`
1. Delete docker volume: `docker volume rm external-services_hash-postgres-data` (alternative `docker volume prune` if you're okay with wiping _every_ docker volume not in use) 
1. Rebuild environment:  `yarn external-services up --build` (and any other docker compose arg if you wish, e.g. `--detach` if that's desired.)

**When running at the`/packages/graph` folder**:
1. Change directory to `/packages/graph/deployment`, stop container: `docker compose --env-file ../.env down`
1. Delete docker volume: `docker volume rm deployment_hash-graph-data` (again as above, you can prune instead)
1. Rebuild environment: `docker compose --env-file ../.env up --build` (and any other desired args..)
- After this, you can use `cargo make deployment-up` and `cargo make deployment-up` as usual.

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1201095311341924/1202681321906843/f) _(internal)_


## 🚫 Blocked by





## 🔍 What does this change?




- Renames `inactivate` to `remove` for links
- Introduces a new table `link_histories` table that store historic links
    -  The existing `links` table will contain `current` links which do not have a `removed_at` timestamp
    -  The historic table will contain provenance metadata about who/when the removal happened for the link, and allow us to look at links for any point in time.
- Removes irrelevant structs for links
- Removes `is_active` from `LinkRecord` and link query resolver
- Adds ordered results for `link` reads

## 📜 Does this require a change to the docs?



- OpenAPI spec has been updated

## ⚠️ Known issues


- Still no way to create ordered links, as we don't require it just yet
- The changes here break existing dev environments

## 🐾 Next steps


-  Allow creating ordered links [Asana task](https://app.asana.com/0/1202805690238892/1202937382769278/f) _(internal)_
   - Potentially fractional link indexing, but further down the line [Asana task](https://app.asana.com/0/1200211978612931/1202085856561975/f) _(internal)_

## 🛡 What tests cover this?



- Existing integration tests

## ❓ How to test this?



1.  Checkout the branch / view the deployment
1.  Run HTTP tests
1.  Or run jest integration tests for links:
    1.  Recreate and migrate DB and run backend
         -  Follow steps in the top of this description to recreate env, you should have running external services
    1.  In another terminal, from the `packages/hash/integration` path run `yarn jest "knowledge/link"`
    1.  Confirm all tests pass
## 📹 Demo
